lib/Mojo/WebSocketProxy/Dispatcher.pm

```diff
@@ -66,40 +72,68 @@ sub on_message {
 
     my $config = $c->wsp_config->{config};
 
-    my $result;
     my $req_storage = {};
     $req_storage->{args} = $args;
-    timeout 15 => sub {
-        $result = Mojo::WebSocketProxy::Parser::parse_req($c, $req_storage);
-        if (!$result
-            && (my $action = $c->dispatch($args)))
-        {
-            %$req_storage = (%$req_storage, %$action);
-            $req_storage->{method} = $req_storage->{name};
-            $result = $c->before_forward($req_storage);
-
-            # Don't forward call to RPC if any before_forward hook returns response
-            # Or if there is instead_of_forward action
-            unless ($result) {
-                $result =
-                      $req_storage->{instead_of_forward}
-                    ? $req_storage->{instead_of_forward}->($c, $req_storage)
-                    : $c->forward($req_storage);
-            }
-
-            $c->after_forward($result, $req_storage);
-        } elsif (!$result) {
-            $c->app->log->debug("unrecognised request: " . $c->dumper($args));
-            $result = $c->wsp_error('error', 'UnrecognisedRequest', 'Unrecognised request.');
+
+    my $result = Mojo::WebSocketProxy::Parser::parse_req($c, $req_storage);
+
+    my $result_f = $result ? Future->fail($result) : Future->done;
+
+    # main processing pipeline
+    my $f = $result_f->then(
+        sub {
+            my $action = $c->dispatch($args);
+            Future->fail($result = $c->wsp_error('error', 'UnrecognisedRequest', 'Unrecognised request'))
+                unless $action;
+            Future->done($action);
         }
-    };
-    if ($@) {
-        warn("$$ timeout for " . JSON::to_json($args));
-    }
+        )->then(
+        sub {
+            my $action = shift;
 
-    $c->send({json => $result}, $req_storage) if $result;
+            %$req_storage = (%$req_storage, %$action);
+            $req_storage->{method} = $req_storage->{name};
 
-    $c->_run_hooks($config->{after_dispatch} || []);
+            my $f = $c->before_forward($req_storage)->then(
+                sub {
+                    my $next =
+                        $req_storage->{instead_of_forward}
+                        ? sub { $req_storage->{instead_of_forward}->($c, $req_storage) }
+                        : sub { $c->forward($req_storage) };
+                    Future->done($next->());
+                }
+                )->else(
+                sub {
+                    $result = shift;
```

Comment:
It is the same `$result` which is defined in line #78. Yeah, that's silly, but that's original logic, and I was asked to keep backward compatibility.